### ISSUE TITLE
Mirage-Firewall Booting Error

### ISSUE THREAD

**Issue by bluesteal** [mirage/qubes-mirage-firewall#177]
Getting an error when trying to boot mirage-firewall; it launches at first, and then it crashes due to an error. The log file shows 
the following error:

Fatal error: exception Xs_protocol.Error("EACCES")
Raised at Xs_protocol.response in file "dunivers/ocaml-xenstore/client_lwt/xs_client_lwt.ml", line 318, characters 13-50
Called from Lwt.Sequential_composition.bind.create_result_promise_and_callback_if_deferred.callback in file "duniverse/lwt/src/core/lwt.ml", line 1849
Solo5: solo5_exit(2) called

**Comment by palainp** [comment]
Hi, thanks for the report, I think this issue is strongly related to #107.

Could it be possible for you to compile a test-case mirage-firewall (as described in https://forum.qubes-os.org/t/internal-error-libxenlight/18891/30)? If you can't I can provide you one I compiled myself.

This will print more lines where I suspect a possible race condition with Xen. And could you provide a full run log and not only the last failing part, as well as some information on your setup (what have you changed recently, what is the system doing when this occurs, etc.) Many thanks for helping solving this issue :)

**Comment by bluesteal** [comment]
The only change that was made recently was upgrading to fedora-37 as a base template.

**Comment by bluesteal** [comment]
In reference to https://github.com/mirage/qubes-mirage-firewall/issues/107
There is a comment regarding a permission-denied error from xenstore.

**Comment by bluesteal** [comment]
Ignored unknown IPv4 message from uplink: Ignoring non-TCP/UDP packet

It seems to me to be an attempt at opening a network connection that is getting a permission denied-error from xenstore
for some reason. But there has been no change in permissions at all so I'm not sure what actually has changed here.

**Comment by palainp** [comment]
It would really help if you compile the unikernel with the commands described in the Qubes forum (link above), and if you verbatim copy the logs (not some isolated portions but every lines from the solo5 banner to the exception failure).

**Comment by bluesteal** [comment]
where is dunivers/ocaml-xenstore/client_lwt/xs_client_lwt.ml located? Is it in dom0? If mirage-firewall cannot be booted, can this directory be accessed via dom0?

**Comment by bluesteal** [comment]
yes; I just couldn't recall the com0 directory off the top of my head.

**Comment by palainp** [comment]
Thanks this will probably gives some additional informations :)
And BTW, all compile commands should be run into an AppVM, not dom0.

**Comment by bluesteal** [comment]
The directory used to copy the unikernel into was /var/lib/qubes/vm-kernels

**Comment by bluesteal** [comment]
but I don't see dunivers/ocaml-xenstore/client_lwt/xs_client_lwt.ml in this directory

**Comment by palainp** [comment]
You should strictly follow the commands given on the Qubes forum, in an AppVM:
- download opam 2.1
- init opam and select Ocaml 4.14.1 with `opam switch`
- pin a specific version for helping the débug task
- clone qubes-mirage-fw, this will create a directory where all the following have to be run from
- compile the unikernel 
- upload `dist/qubes-firewall.xen` to dom0 in the vm-kernels directory

The .ml file is just a source file and never was in dom0. Dont bother with it.

**Comment by bluesteal** [comment]
I understand your instructions and I appreciate your response; what I am trying to get at is the root cause of this issue and 
why it is all of a sudden occurring. Is this because there are library dependencies in the unikernel that are out of date with respect to fedora-37?

**Comment by palainp** [comment]
If that's a race condition when the unikernel is connecting to uplink, it's hard to say what is the root cause and why it appears now (Just curiosity, does it behave normally if you switch back to f36?)
It can be less CPU usage that leaves the unikernel with more ressources, it can be a slightly slower/faster AppVM when using f37, I dont really know and never been able to reproduce. So I'm glad you can do some tests :)

**Comment by bluesteal** [comment]
I'm going to try reverting to f36 actually

**Comment by palainp** [comment]
Ok, something else had changed, but never mind.

In case you want to try it, I compiled an unikernel that includes more log traces when connecting to its uplink, you can find it at https://filebin.net/qdbih3uk978ruber (the sha256sum is e5e5fbbe0fb5007524c67943335589a6fb0e317979ef092f4929cadc78e53361).

**Comment by bluesteal** [comment]
Is there any way to increase the verbosity of the logs without having to install a new unikernel?

**Comment by palainp** [comment]
Unfortunately not, this build specifically adds some new logs lines. In dom0, you can save your actual unikernel, and copy it back once you had run this one and extracted the logs.

**Comment by bluesteal** [comment]
There must have been a change to xenstore then, which is dom0 code. If the mirage-firewall doesn't receive any updates; then the update must've occurred within dom0. So this is really a Qubes dom0 repo issue

**Comment by palainp** [comment]
No, with the log lines, I also added some instructions for mirage-fw to wait the other side of its backend (if you want some details, in Ocaml : https://github.com/mirage/mirage-net-xen/compare/main...palainp:mirage-net-xen:wait_backend). This could bé something done only on mirage-fw side.

If that works with this unikernel, it means that it's really a race condition, mirage-fw tries to read on xenstore to early. If not, the issue is elsewhere and more investigations have to be done.

**Comment by bluesteal** [comment]
Is there any way to just make the code modification in place?

**Comment by palainp** [comment]
Not at all, Qubes uses a lot of python scripts but here the unikernel is compiled and is a binary file. The best is to save your existing unikernel (in case you want to get it back, but as I understand the issue it currently fails) and put that test version in place (it should behave exactly the same as the original one with a bit more log lines at startup and hopefully could fixes your issue).

**Comment by bluesteal** [comment]
This issue must be originating from within the dom0 back end. The only recent change has been a dom0 update.

**Comment by palainp** [comment]
It could be (but site could also be an improvement on Qubes side that trigger more easily the race condition). From my point of view, if there is a race condition in the mirage-fw code it worth to identify fix it :)
As you can reproduce it (and I can't) I hope you won't mind to test that out :)

**Comment by bluesteal** [comment]
I think the issue is within the xen backend as this is the only recent change to the system.

**Comment by palainp** [comment]
Well, it's hard to say without testing and anyway the issue can still be in the mirage ecosystem code.

At this stage it will really helps if you can copy the compiled unikernel as you did for the original one with something, in dom0, similar to `qvm-run -p dev 'cat mirage-firewall/vmlinuz' > vmlinuz` and verbatim copy the logs.

**Comment by bluesteal** [comment]
I filed a bug with the Qubes project.

**Comment by palainp** [comment]
That's great, but as the root cause still hasn't been clearly identified, and the error seems to be restricted to mirage unikernels I suppose that there's something to update on our side (I guess Qubes devs could legitimately reply that :) ). 

I suppose something in `mirage-net-xen` or around has to be fixed as the code reporting the exception comes from there. Alas as I'm not able to trigger this exception by myself and test code updates, I'm desperately looking for someone who can reliably trigger it and test code updates :)

**Comment by bluesteal** [comment]
mirage-net-xen is this mirageOS code or QubesOS code?

**Comment by palainp** [comment]
It's mirageOS :)

**Comment by bluesteal** [comment]
But what could possibly have changed? It's not as though the binary could have been updated...it is very strange.

**Comment by palainp** [comment]
Yes this is strange and hard to tell what is the root cause. I suppose this could be seen as a side effect from an update but I really don't know.

For me the main idea is to rule out an possible race condition in our code while thé issue is reproducible :)

**Comment by bluesteal** [comment]
The race condition can only be mitigated with an updated unikernel?

**Comment by palainp** [comment]
Yes if it's due to some code in any mirageOS part.

**Comment by palainp** [comment]
In the case you didn't get the unikernel before end of life for the previous link, here's a new link (same sha256sum) : https://filebin.net/jbv8d7hcasxhg0hc

**Comment by bluesteal** [comment]
Could this problem be related to cryptographic keys?

**Comment by bluesteal** [comment]
My suspicion is that there is an issue with a Xen API that generates entropy for random number generation (RNG) for the purpose of generating cryptographic keys, or some other purpose. The reason is because there are continual crashes with the following error:

RSOD (Red Screen Of Death) and an error message:

systemd uninitialized urandom read crash

In Unix-like operating systems, /dev/random and /dev/urandom are special files that serve as cryptographically secure pseudorandom number generators. They allow access to environmental noise collected from device drivers and other sources.[1] /dev/random typically blocked if there was less entropy available than requested; more recently (see below for the differences between operating systems) it usually blocks at startup until sufficient entropy has been gathered, then unblocks permanently. The /dev/urandom device typically was never a blocking device, even if the pseudorandom number generator seed was not fully initialized with entropy since boot. Not all operating systems implement the same methods for /dev/random and /dev/urandom.

**Comment by xaki23** [comment]
i am seeing this EACCES crash-on-start here now. 
mfw build from 2022-12-11 that has been working fine until "just now".
only thing that comes to mind is a qubes-core-* update some hours ago.
doesnt "feel" like a race condition either. (cant get it to work no matter how many restarts i try, in different configurations. didnt try whole-system reboot yet. so either the dom0 update changed something(tm) or something is not running (properly) because of the update.)

@palainp should i read through all this and try your experimental build or try something else?

**Comment by palainp** [comment]
Yes I'd be very glad if you can try it out :)

You should have more verbosity around the faulty xenstore read access with the mirage-net-xen branch, so you just have to pin it and compile qubes-mirage-fw with the current head.

**Comment by xaki23** [comment]
> Yes I'd be very glad if you can try it out :)

(sent you a mail @enssat about how to _probably_ get down on this a lot faster)

**Comment by palainp** [comment]
"Good" news, I'm not able to fully reproduce this issue on my laptop (so something in `python3-qubesadmin-4.1.31-1.fc32.noarch, qubes-core-admin-client-4.1.31-1.fc32.noarch,  qubes-core-dom0-4.1.31-1.fc32.noarch                                 qubes-core-qrexec-4.1.22-1.fc32.x86_64, qubes-core-qrexec-dom0-4.1.22-1.fc32.x86_64, qubes-core-qrexec-libs-4.1.22-1.fc32.x86_64, qubes-mgmt-salt-dom0-update-4.1.11-1.fc32.noarch` had changed something :) ).

Unfortunately it will be hard/long to investigate. I've follow the issue to https://github.com/mirage/mirage-net-xen/blob/939dfcbfc31f6c396532c08d8857f2672d90f2f3/lib/frontend.ml#L212 and I'm not sure how to add debug traces in https://github.com/mirage/shared-memory-ring/blob/7e6a5e8a78eb62552d6519ad4c00d4d65664e65e/lwt/lwt_ring.ml#L59 .

To my understanding what's going on is:
- we have to deal something with arp (https://github.com/mirage/qubes-mirage-firewall/blob/609f5295c7b315886244426b685807244c7dbe81/uplink.ml#L56 and then https://github.com/mirage/ethernet/blob/f566a7135d481302637587b566430cd66698d630/src/ethernet.ml#L83)
- the arp code (https://github.com/mirage/arp/blob/acd52c1e6ea17eb81b82bb06cb9847aebca59763/mirage/arp.ml#L76) have `out=None` and `wake=Some (mac, (_, u))`
- a task is awaken with a resolved mac address and fails to write on the xenstore ring.

**Comment by palainp** [comment]
Thank you for your reply!
The packet is a direct request from our uplink (based on our gratuitous arp sended at startup). I managed to move a step forward by adding `Arp.query arp gateway >>= fun _ ->` in uplink.ml:connect, but it still fails somewhere after, this is probably not useful :(

The best I can do to describe the error is by adding a log line at:
https://github.com/mirage/mirage-net-xen/blob/939dfcbfc31f6c396532c08d8857f2672d90f2f3/lib/frontend.ml#L214
which will call https://github.com/mirage/shared-memory-ring/blob/7e6a5e8a78eb62552d6519ad4c00d4d65664e65e/lwt/lwt_ring.ml#L59
And I have the following:
```
[2023-07-04 14:31:29] 2023-07-04 12:31:29 -00:00: WRN [net-xen frontend] from PA in tx_poll
[2023-07-04 14:31:29] Fatal error: exception Xs_protocol.Error("EACCES")
[2023-07-04 14:31:29] Raised at Xs_protocol.response in file "duniverse/ocaml-xenstore/core/xs_protocol.ml", line 685, characters 13-28
[2023-07-04 14:31:29] Called from Xs_client_lwt.Client.rpc.(fun) in file "duniverse/ocaml-xenstore/client_lwt/xs_client_lwt.ml", line 318, characters 13-50
[2023-07-04 14:31:29] Called from Lwt.Sequential_composition.bind.create_result_promise_and_callback_if_deferred.callback in file "duniverse/lwt/src/core/lwt.ml", line 1849, characters 23-26
[2023-07-04 14:31:29] Solo5: solo5_exit(2) called
```
Of course I can see some `from PA in tx_poll` before so to me the Xen ringbuffer code is ok, so I'm not sure how to proceed to move forward from here. It's like tx_poll have no pre-reserved slot in the buffer, but I don't understand why it appears now :(

**Comment by palainp** [comment]
I'll try to look on the differences. Here are the old versions (except kernels packages):
```
  Verifying        : python3-qubesadmin-4.1.31-1.fc32.noarch                                                                      5/20 
  Verifying        : python3-qubesadmin-4.1.27-1.fc32.noarch                                                                      6/20 
  Verifying        : qubes-core-admin-client-4.1.31-1.fc32.noarch                                                                 7/20 
  Verifying        : qubes-core-admin-client-4.1.27-1.fc32.noarch                                                                 8/20 
  Verifying        : qubes-core-dom0-4.1.31-1.fc32.noarch                                                                         9/20 
  Verifying        : qubes-core-dom0-4.1.30-1.fc32.noarch                                                                        10/20 
  Verifying        : qubes-core-qrexec-4.1.22-1.fc32.x86_64                                                                      11/20 
  Verifying        : qubes-core-qrexec-4.1.21-1.fc32.x86_64                                                                      12/20 
  Verifying        : qubes-core-qrexec-dom0-4.1.22-1.fc32.x86_64                                                                 13/20 
  Verifying        : qubes-core-qrexec-dom0-4.1.21-1.fc32.x86_64                                                                 14/20 
  Verifying        : qubes-core-qrexec-libs-4.1.22-1.fc32.x86_64                                                                 15/20 
  Verifying        : qubes-core-qrexec-libs-4.1.21-1.fc32.x86_64                                                                 16/20 
  Verifying        : qubes-mgmt-salt-dom0-update-4.1.11-1.fc32.noarch                                                            17/20 
  Verifying        : qubes-mgmt-salt-dom0-update-4.1.10-1.fc32.noarch                                                            18/20
```

**Comment by palainp** [comment]
It's very peculiar and that doesn't means we out of troubles, but while doing recursive diffs between packages (no changelog provided :( ), I spotted a lot of changes in qmemman and this led me to try using #176 (in fact a backported version), which seems to work as is. @xaki23 would you mind to try out with the "no memory report" PR?

**Comment by xaki23** [comment]
building from no-memreport works (using qubes-builder), 
the built mfw starts (so solves this issue), 
and the VM passes initial functional tests. 
:+1:

**Comment by palainp** [comment]
Fixed by #176.
